File: P03_WumpusWorld/P03_WumpusWorld/client/kb.go
// kb.go
//
// Knowledge base for the Wumpus World expressed as CNF clauses. The
// knowledge base is queried through the DPLL engine defined in dpll.go to
// decide whether a given cell is safe (i.e. free of both pits and the
// Wumpus) and, when possible, to pinpoint the Wumpus location.
//
// Propositional variables.
//   Pit(x, y)     : there is a pit at (x, y)
//   Wumpus(x, y)  : the Wumpus is at (x, y)
//
// Variable identifiers are assigned as follows for a 4×4 grid:
//   Pit(x, y)     = y*GridSize + x + 1           (range 1 … 16)
//   Wumpus(x, y)  = GridSize*GridSize + y*GridSize + x + 1   (range 17 … 32)
package main

// GridSize is the side length of the square Wumpus World grid. This must
// match the value used by the server (see game.go).
const GridSize = 4

// pitVar returns the variable id representing "there is a pit at (x, y)".
func pitVar(x, y int) int {
	return y*GridSize + x + 1
}

// wumpusVar returns the variable id representing "the Wumpus is at (x, y)".
func wumpusVar(x, y int) int {
	return GridSize*GridSize + y*GridSize + x + 1
}

// pLit builds a positive literal for variable v.
func pLit(v int) Literal { return Literal(v) }

// nLit builds a negative literal for variable v.
func nLit(v int) Literal { return Literal(-v) }

// neighborsOf returns the valid orthogonal neighbors of (x, y).
func neighborsOf(x, y int) [][2]int {
	deltas := [4][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
	result := make([][2]int, 0, 4)
	for _, d := range deltas {
		nx, ny := x+d[0], y+d[1]
		if nx >= 0 && nx < GridSize && ny >= 0 && ny < GridSize {
			result = append(result, [2]int{nx, ny})
		}
	}
	return result
}

// KB is the propositional knowledge base. It stores CNF clauses that encode
// everything the agent currently knows about the world.
type KB struct {
	clauses CNF
}

// NewKB constructs a fresh knowledge base seeded with:
//   - the starting cell (0,0) is free of pits and the Wumpus;
//   - at most one Wumpus exists on the grid.
//
// The "at most one Wumpus" constraint is what enables DPLL to locate the
// Wumpus by elimination, which in turn unlocks shooting plans.
func NewKB() *KB {
	kb := &KB{}

	// The agent starts at (0, 0) and is alive, so the start cell is safe.
	kb.clauses = append(kb.clauses, Clause{nLit(pitVar(0, 0))})
	kb.clauses = append(kb.clauses, Clause{nLit(wumpusVar(0, 0))})

	// At most one Wumpus: for each pair of cells (i, j), ¬W(i) ∨ ¬W(j).
	total := GridSize * GridSize
	for i := 0; i < total; i++ {
		wi := GridSize*GridSize + i + 1
		for j := i + 1; j < total; j++ {
			wj := GridSize*GridSize + j + 1
			kb.clauses = append(kb.clauses, Clause{nLit(wi), nLit(wj)})
		}
	}

	// At least one Wumpus somewhere on the board. Strictly speaking, the
	// game always spawns a Wumpus, so we add this to help DPLL pin down its
	// location once enough stench/no-stench evidence has been gathered.
	var wLits Clause
	for x := 0; x < GridSize; x++ {
		for y := 0; y < GridSize; y++ {
			wLits = append(wLits, pLit(wumpusVar(x, y)))
		}
	}
	kb.clauses = append(kb.clauses, wLits)

	return kb
}

// AddVisitObservation records that the agent entered (x, y) and survived
// (so no pit and no live Wumpus there) and reports perceiving the given
// stench / breeze flags at that cell.
func (kb *KB) AddVisitObservation(x, y int, stench, breeze bool) {
	// The current cell is safe because the agent is alive on it.
	kb.clauses = append(kb.clauses, Clause{nLit(pitVar(x, y))})
	kb.clauses = append(kb.clauses, Clause{nLit(wumpusVar(x, y))})

	nb := neighborsOf(x, y)

	// Stench ↔ some neighbor has the Wumpus.
	if stench {
		// Stench: at least one neighbor has the Wumpus.
		disj := make(Clause, 0, len(nb))
		for _, n := range nb {
			disj = append(disj, pLit(wumpusVar(n[0], n[1])))
		}
		if len(disj) > 0 {
			kb.clauses = append(kb.clauses, disj)
		}
	} else {
		// No stench: no neighbor has the Wumpus.
		for _, n := range nb {
			kb.clauses = append(kb.clauses, Clause{nLit(wumpusVar(n[0], n[1]))})
		}
	}

	// Breeze ↔ some neighbor has a pit.
	if breeze {
		disj := make(Clause, 0, len(nb))
		for _, n := range nb {
			disj = append(disj, pLit(pitVar(n[0], n[1])))
		}
		if len(disj) > 0 {
			kb.clauses = append(kb.clauses, disj)
		}
	} else {
		for _, n := range nb {
			kb.clauses = append(kb.clauses, Clause{nLit(pitVar(n[0], n[1]))})
		}
	}
}

// KillWumpus registers that the Wumpus has been killed (a scream was
// heard): for every cell (x, y) we assert ¬Wumpus(x, y). Any cell that was
// previously unsafe only because of the Wumpus can now be deduced safe.
func (kb *KB) KillWumpus() {
	for x := 0; x < GridSize; x++ {
		for y := 0; y < GridSize; y++ {
			kb.clauses = append(kb.clauses, Clause{nLit(wumpusVar(x, y))})
		}
	}
}

// IsSafe returns true when the knowledge base can prove that (x, y)
// contains neither a pit nor a live Wumpus.
func (kb *KB) IsSafe(x, y int) bool {
	safeFromPit := Entails(kb.clauses, nLit(pitVar(x, y)))
	if !safeFromPit {
		return false
	}
	return Entails(kb.clauses, nLit(wumpusVar(x, y)))
}

// HasConfirmedWumpus returns (x, y, true) if DPLL proves the Wumpus must
// live at a single specific cell. Otherwise (0, 0, false) is returned.
func (kb *KB) HasConfirmedWumpus() (int, int, bool) {
	for x := 0; x < GridSize; x++ {
		for y := 0; y < GridSize; y++ {
			if Entails(kb.clauses, pLit(wumpusVar(x, y))) {
				return x, y, true
			}
		}
	}
	return 0, 0, false
}

// MaybePit reports whether it is consistent with the KB that (x, y) holds a
// pit. Useful for ordering risky moves when no provably safe move exists.
func (kb *KB) MaybePit(x, y int) bool {
	// It is possible that (x, y) has a pit iff KB ∧ Pit(x, y) is SAT.
	test := make(CNF, len(kb.clauses)+1)
	copy(test, kb.clauses)
	test[len(kb.clauses)] = Clause{pLit(pitVar(x, y))}
	return IsSAT(test)
}

File: P03_WumpusWorld/P03_WumpusWorld/python_client/kb.py
"""
kb.py
=====
Propositional Knowledge Base for the Wumpus World expressed as CNF clauses.
The KB is queried through the DPLL engine (dpll.py) to decide whether a given
cell is safe (free of pits and the Wumpus) and, when possible, to pinpoint the
Wumpus location.

Propositional variables (4×4 grid):
  pit_var(x, y)     = y * GRID_SIZE + x + 1            → range  1 …16
  wumpus_var(x, y)  = GRID_SIZE² + y * GRID_SIZE + x + 1 → range 17 …32

Encoding matches the Go implementation in client/kb.go exactly.
"""

from __future__ import annotations
from typing import List, Tuple

from dpll import CNF, Clause, Literal, is_sat, entails

GRID_SIZE: int = 4   # must match the server value


# ---------------------------------------------------------------------------
# Variable helpers
# ---------------------------------------------------------------------------

def pit_var(x: int, y: int) -> int:
    """Variable id for 'there is a pit at (x, y)'."""
    return y * GRID_SIZE + x + 1


def wumpus_var(x: int, y: int) -> int:
    """Variable id for 'the Wumpus is at (x, y)'."""
    return GRID_SIZE * GRID_SIZE + y * GRID_SIZE + x + 1


def p_lit(v: int) -> Literal:
    """Positive (TRUE) literal for variable v."""
    return v


def n_lit(v: int) -> Literal:
    """Negative (FALSE) literal for variable v."""
    return -v


def neighbors_of(x: int, y: int) -> List[Tuple[int, int]]:
    """Return the valid orthogonal neighbors of (x, y) within the grid."""
    result = []
    for dx, dy in ((-1, 0), (1, 0), (0, -1), (0, 1)):
        nx, ny = x + dx, y + dy
        if 0 <= nx < GRID_SIZE and 0 <= ny < GRID_SIZE:
            result.append((nx, ny))
    return result


# ---------------------------------------------------------------------------
# Knowledge Base
# ---------------------------------------------------------------------------

class KB:
    """
    Propositional Knowledge Base that accumulates CNF clauses derived from
    the agent's perceptions and uses DPLL for inference.
    """

    def __init__(self) -> None:
        self.clauses: CNF = []
        self._seed()

    # ------------------------------------------------------------------
    # Initialisation
    # ------------------------------------------------------------------

    def _seed(self) -> None:
        """
        Seed the KB with world axioms:
          • (0,0) is safe – the agent starts there alive.
          • At most one Wumpus exists on the board
            (¬W(i) ∨ ¬W(j) for every pair i ≠ j).
          • At least one Wumpus exists (disjunction of all W(x,y)).
        """
        # Start cell is safe
        self.clauses.append([n_lit(pit_var(0, 0))])
        self.clauses.append([n_lit(wumpus_var(0, 0))])

        total = GRID_SIZE * GRID_SIZE

        # At most one Wumpus: ¬W(i) ∨ ¬W(j) for every pair i < j
        for i in range(total):
            wi = GRID_SIZE * GRID_SIZE + i + 1
            for j in range(i + 1, total):
                wj = GRID_SIZE * GRID_SIZE + j + 1
                self.clauses.append([n_lit(wi), n_lit(wj)])

        # At least one Wumpus somewhere on the board
        w_lits: Clause = []
        for x in range(GRID_SIZE):
            for y in range(GRID_SIZE):
                w_lits.append(p_lit(wumpus_var(x, y)))
        self.clauses.append(w_lits)

    # ------------------------------------------------------------------
    # Observation recording
    # ------------------------------------------------------------------

    def add_visit_observation(
        self, x: int, y: int, stench: bool, breeze: bool
    ) -> None:
        """
        Record that the agent entered (x, y) and survived (no pit, no live
        Wumpus there), then add stench/breeze evidence.

        Stench ↔ at least one neighbour has the Wumpus.
        Breeze ↔ at least one neighbour has a pit.
        """
        # Survived → cell is safe
        self.clauses.append([n_lit(pit_var(x, y))])
        self.clauses.append([n_lit(wumpus_var(x, y))])

        nb = neighbors_of(x, y)

        # ---- Stench ----
        if stench:
            # ∃ neighbour with Wumpus
            disj: Clause = [p_lit(wumpus_var(nx, ny)) for nx, ny in nb]
            if disj:
                self.clauses.append(disj)
        else:
            # ∀ neighbours: no Wumpus
            for nx, ny in nb:
                self.clauses.append([n_lit(wumpus_var(nx, ny))])

        # ---- Breeze ----
        if breeze:
            # ∃ neighbour with pit
            disj = [p_lit(pit_var(nx, ny)) for nx, ny in nb]
            if disj:
                self.clauses.append(disj)
        else:
            # ∀ neighbours: no pit
            for nx, ny in nb:
                self.clauses.append([n_lit(pit_var(nx, ny))])

    def kill_wumpus(self) -> None:
        """
        Register that the Wumpus has been killed (scream perceived).
        Asserts ¬Wumpus(x, y) for every cell so that DPLL can now prove
        wumpus-related cells safe.
        """
        for x in range(GRID_SIZE):
            for y in range(GRID_SIZE):
                self.clauses.append([n_lit(wumpus_var(x, y))])

    # ------------------------------------------------------------------
    # Inference queries
    # ------------------------------------------------------------------

    def is_safe(self, x: int, y: int) -> bool:
        """
        Return True iff KB proves (x, y) is free of both pits and the Wumpus.

        Uses DPLL refutation:
            KB ⊨ ¬Pit(x,y)   AND   KB ⊨ ¬Wumpus(x,y)
        """
        if not entails(self.clauses, n_lit(pit_var(x, y))):
            return False
        return entails(self.clauses, n_lit(wumpus_var(x, y)))

    def has_confirmed_wumpus(self) -> Tuple[int, int, bool]:
        """
        Return (wx, wy, True) if DPLL proves the Wumpus must be at exactly
        one cell.  Otherwise returns (0, 0, False).
        """
        for x in range(GRID_SIZE):
            for y in range(GRID_SIZE):
                if entails(self.clauses, p_lit(wumpus_var(x, y))):
                    return x, y, True
        return 0, 0, False

    def confirmed_or_only_wumpus(self) -> Tuple[int, int, bool]:
        """
        Extended wumpus location query used by the agent to decide when to shoot.

        Strategy (two levels of confidence):

        Level 1 — Full certainty (DPLL refutation):
            KB ⊨ W(x,y)  iff  KB ∧ ¬W(x,y)  is UNSAT.
            If this holds for some cell, the Wumpus MUST be there.

        Level 2 — Only consistent candidate (SAT check):
            For every cell (x,y), test whether KB ∧ [W(x,y)] is SAT.
            If exactly ONE cell passes the test, the Wumpus can only be
            there given all current evidence → safe to shoot.

        CNF query for "Is (x,y) a consistent Wumpus location?":
            test_clauses = self.clauses + [[wumpus_var(x, y)]]
            is_sat(test_clauses)   # SAT → consistent · UNSAT → ruled out

        Returns (wx, wy, True) on success, (0, 0, False) otherwise.
        """
        # --- Level 1: full logical certainty ---
        for x in range(GRID_SIZE):
            for y in range(GRID_SIZE):
                if entails(self.clauses, p_lit(wumpus_var(x, y))):
                    return x, y, True

        # --- Level 2: unique consistent candidate ---
        candidates = []
        for x in range(GRID_SIZE):
            for y in range(GRID_SIZE):
                # CNF query: assume W(x,y)=True as a unit clause, check SAT
                test = [list(c) for c in self.clauses]
                test.append([p_lit(wumpus_var(x, y))])
                if is_sat(test):
                    candidates.append((x, y))

        if len(candidates) == 1:
            wx, wy = candidates[0]
            return wx, wy, True

        return 0, 0, False

    def maybe_pit(self, x: int, y: int) -> bool:
        """
        Return True if it is consistent with the KB that (x, y) contains a
        pit.  Useful for risk-ordering moves when no safe frontier exists.
        """
        test = [list(c) for c in self.clauses]
        test.append([p_lit(pit_var(x, y))])
        return is_sat(test)
